### DISCUSSION TITLE
Help with examples compilation

### DISCUSSION THREAD

**Discussion by rodion-maltsev** [hathach/tinyusb#1030] [Q&A] [answered]
I need some help with examples compilation.
Im trying to do this on Windows 10 x64 and Windows 7 x64 machines.
I know nothing at all and do not understand in the linux system usage.
My task is to get source code of cdc_msc and cdc_msc_freertos examples to use it on stm32h753 with IAR EWB for ARM.

Here is a list of tools I used and errors I received:
**Chocolatey.**
When I dont specify SHELL variable:
```
c:\tinyusb\examples\device\cdc_msc>C:\ProgramData\chocolatey\bin\make BOARD=stm32h743nucleo all
/usr/bin/sh: -c: line 0: syntax error near unexpected token `('
/usr/bin/sh: -c: line 0: `for %%i in ( ../../.. ) do echo %%~fi'
You must provide a BOARD parameter with 'BOARD='
../../make.mk:36: *** Invalid BOARD specified.  Stop.
```
So i specify the SHELL, but getting this error:
```
c:\tinyusb\examples\device\cdc_msc>C:\ProgramData\chocolatey\bin\make SHELL=cmd.exe BOARD=stm32h743nucleo all
"PORT0 Full Speed"
AS startup_stm32h743xx_asm.o
process_begin: CreateProcess(NULL, arm-none-eabi-gcc -x assembler-with-cpp -DSTM32H743xx -DHSE_VALUE=8000000 -flto -mthumb -mabi=aapcs -mcpu=cortex-m7 -mfloat-abi=hard -mfpu=fpv5-d16 -nostdlib -nostartfiles -DCFG_TUSB_MCU=OPT_MCU_STM32H7 -DBOARD_DEVICE_RHPORT_NUM=0 -Wno-error=maybe-uninitialized -Wno-error=cast-align -ggdb -fdata-sections -ffunction-sections -fsingle-precision-constant -fno-strict-aliasing -Wdouble-promotion -Wstrict-prototypes -Wstrict-overflow -Wall -Wextra -Werror -Wfatal-errors -Werror-implicit-function-declaration -Wfloat-equal -Wundef -Wshadow -Wwrite-strings -Wsign-compare -Wmissing-format-attribute -Wunreachable-code -Wcast-align -Wcast-function-type -Os -Ic:/tinyusb/hw/bsp/stm32h7/boards/stm32h743nucleo -Ic:/tinyusb/lib/CMSIS_5/CMSIS/Core/Include -Ic:/tinyusb/hw/mcu/st/cmsis_device_h7/Include -Ic:/tinyusb/hw/mcu/st/stm32h7xx_hal_driver/Inc -Ic:/tinyusb/hw/bsp/stm32h7 -Isrc -Ic:/tinyusb/hw -Ic:/tinyusb/src -c -o _build/stm32h743nucleo/obj/hw/mcu/st/cmsis_device_h7/Source/Templates/gcc/startup_stm32h743xx_asm.o c:/tinyusb/hw/mcu/st/cmsis_device_h7/Source/Templates/gcc/startup_stm32h743xx.s, ...) failed.
make (e=2): ═х єфрхЄё  эрщЄш єърчрээ√щ Їрщы.
make: *** [../../rules.mk:119: _build/stm32h743nucleo/obj/hw/mcu/st/cmsis_device_h7/Source/Templates/gcc/startup_stm32h743xx_asm.o] Error 2
```
**MINGW**
```
Rodion@DESKTOP-172PB8K MINGW64 /c/tinyusb/examples/device/cdc_msc (master)
$ make BOARD=stm32h743nucleo all
"PORT0 Full Speed"
../../rules.mk:73: *** multiple target patterns.  Stop.
```
I searched for this problem in web, most common issue is spaces in makefile or in paths to directories/files, but there are none.

**Cygwin**
```
Rodion@DESKTOP-172PB8K /cygdrive/c/tinyusb/examples/device/cdc_msc
$ make BOARD=stm32h743nucleo all
"PORT0 Full Speed"
AS startup_stm32h743xx_asm.o
make: arm-none-eabi-gcc: No such file or directory
make: *** [../../rules.mk:119: _build/stm32h743nucleo/obj/hw/mcu/st/cmsis_device_h7/Source/Templates/gcc/startup_stm32h743xx_asm.o] Error 127
```
After arm-none-eabi-gcc setup:
```
Rodion@DESKTOP-172PB8K /cygdrive/c/tinyusb/examples/device/cdc_msc
$ make BOARD=stm32h743nucleo
"PORT0 Full Speed"
AS startup_stm32h743xx_asm.o
arm-none-eabi-gcc.exe: error: /cygdrive/c/tinyusb/hw/mcu/st/cmsis_device_h7/Source/Templates/gcc/startup_stm32h743xx.s: No such file or directory
arm-none-eabi-gcc.exe: fatal error: no input files
compilation terminated.
make: *** [../../rules.mk:119: _build/stm32h743nucleo/obj/hw/mcu/st/cmsis_device_h7/Source/Templates/gcc/startup_stm32h743xx_asm.o] Error 1
```
But I have this file startup_stm32h743xx.s in this directory.

Before compilation I executed all steps from getting_started readme.
And this is a my PATH variable
![image](https://user-images.githubusercontent.com/62758825/129880935-233a4d21-bae4-4240-9c09-1eda426227f1.png)

Can you please help me? Thanks for any replies!

**Comment by hathach** [comment]
this could help https://github.com/hathach/tinyusb/issues/656

**Reply by rodion-maltsev** [reply]
I saw this, but when i specify SHELL variable, on the next step i got 
`process_begin: CreateProcess(NULL, arm-none-eabi-gcc -x assembler-with-cpp -DSTM32H743xx -DHSE_VALUE=8000000 -flto -mthumb -mabi=aapcs -mcpu=cortex-m7 -mfloat-abi=hard -mfpu=fpv5-d16 -nostdlib -nostartfiles -DCFG_TUSB_MCU=OPT_MCU_STM32H7 -DBOARD_DEVICE_RHPORT_NUM=0 -Wno-error=maybe-uninitialized -Wno-error=cast-align -ggdb -fdata-sections -ffunction-sections -fsingle-precision-constant -fno-strict-aliasing -Wdouble-promotion -Wstrict-prototypes -Wstrict-overflow -Wall -Wextra -Werror -Wfatal-errors -Werror-implicit-function-declaration -Wfloat-equal -Wundef -Wshadow -Wwrite-strings -Wsign-compare -Wmissing-format-attribute -Wunreachable-code -Wcast-align -Wcast-function-type -Os -Ic:/tinyusb/hw/bsp/stm32h7/boards/stm32h743nucleo -Ic:/tinyusb/lib/CMSIS_5/CMSIS/Core/Include -Ic:/tinyusb/hw/mcu/st/cmsis_device_h7/Include -Ic:/tinyusb/hw/mcu/st/stm32h7xx_hal_driver/Inc -Ic:/tinyusb/hw/bsp/stm32h7 -Isrc -Ic:/tinyusb/hw -Ic:/tinyusb/src -c -o _build/stm32h743nucleo/obj/hw/mcu/st/cmsis_device_h7/Source/Templates/gcc/startup_stm32h743xx_asm.o c:/tinyusb/hw/mcu/st/cmsis_device_h7/Source/Templates/gcc/startup_stm32h743xx.s, ...) failed.`

**Reply by kkitayam** [reply]
I have reproduced the issue in my Windows 10 environment. I will check the makefiles for the BOARD.

~~We may need to modify makefiles for cmd.exe~~
The above error message indicates the compiler can not find on the PATH environment variable. You can check the toolchain whether exists or not by `where` command on cmd.exe like below.

```
C:\> where arm-none-eabi-gcc
C:\Program Files (x86)\GNU Arm Embedded Toolchain\10 2020-q4-major\bin\arm-none-eabi-gcc.exe
```

In my case, I forgot that `C:\Program Files (x86)\GNU Arm Embedded Toolchain\10 2020-q4-major\bin\gccvar.bat` needs to invoke to set PATH for arm-none-eabi-gcc. After invoking the batch file, succeeded.

**Reply by rodion-maltsev** [reply]
For now, Im succesfully compiled the example under WSL Ubuntu after installing gcc-arm-none-eabi.
Can you pls tell me how I can see the source code of cdc_msc project for stm32h743eval or stm32h743nucleo to remake it for my stm32h753zi. Thanks.

**Reply by rodion-maltsev** [reply]
> I have reproduced the issue in my Windows 10 environment. I will check the makefiles for the BOARD.
> 
> ~We may need to modify makefiles for cmd.exe~
> The above error message indicates the compiler can not find on the PATH environment variable. You can check the toolchain whether exists or not by `where` command on cmd.exe like below.
> 
> ```
> C:\> where arm-none-eabi-gcc
> C:\Program Files (x86)\GNU Arm Embedded Toolchain\10 2020-q4-major\bin\arm-none-eabi-gcc.exe
> ```
> 
> In my case, I forgot that `C:\Program Files (x86)\GNU Arm Embedded Toolchain\10 2020-q4-major\bin\gccvar.bat` needs to invoke to set PATH for arm-none-eabi-gcc. After invoking the batch file, succeeded.

Before invoke, i already havе this variable in PATH, this .bat file adds variable to local user enviroment PATH variable and sure, after invoke gccvar.bat compilation is successful.

**Comment by rodion-maltsev** [comment, marked as answer]
Sure, after invoke gccvar.bat from `C:\Program Files (x86)\GNU Arm Embedded Toolchain\10 2020-q4-major\bin\gccvar.bat` there is a PATH variable for GCC ARM in local user enviroment variables and compilation is successful with chocolatey.

Also you can compile it under WSL.
